examples/face_recognition_11/face_recognition_11/utils.py

```diff
@@ -0,0 +1,188 @@
+            pair = (test_samples[i].locator, test_samples[i].pairs[j].locator)
+            if pair not in seen:
+                scores.append((is_same, similarity))
+                seen.append(pair)
+                seen.append(pair[::-1])
+
+    imposter_scores = sorted(
+        [similarity if similarity is not None else 0.0 for match, similarity in scores if not match],
+        reverse=True,
+    )
+    threshold_idx = int(round(fmr * len(imposter_scores)) - 1)
+    threshold = imposter_scores[threshold_idx] - eps
+    return threshold
+
+
+def compute_baseline_thresholds(
+    test_samples: List[TestSample],
+    ground_truths: List[GroundTruth],
+    inferences: List[Inference],
+    lower_range: int,
+    upper_range: int,
+    num_thresholds: int,
+) -> List[Tuple[float, float]]:
+    baseline_fmr_x = list(np.logspace(lower_range, upper_range, num_thresholds))
+    baseline_thresholds = [compute_threshold(test_samples, ground_truths, inferences, fmr) for fmr in baseline_fmr_x]
+    return zip(baseline_fmr_x, baseline_thresholds)
+
+
+def get_unique_pairs(test_samples: List[TestSample]) -> List[Tuple[str, str]]:
+    pairs = {(ts.locator, pair.locator) for ts in test_samples for pair in ts.pairs}
+    unique_pairs = {(a, b) if a <= b else (b, a) for a, b in pairs}
+    return unique_pairs
+
+
+def compute_pair_metrics(
+    test_samples: List[TestSample],
+    ground_truths: List[GroundTruth],
+    metrics: List[TestSampleMetrics],
+) -> Tuple[list, list, list, list, list, list]:
+    unique_pairs = get_unique_pairs(test_samples)
+    genuine_pairs, imposter_pairs, fm, fnm, pair_failures, fte = {}, {}, {}, {}, {}, {}
+
+    for ts, gt in zip(test_samples, ground_truths):
+        for pair, match in zip(ts.pairs, gt.matches):
+            ab = (ts.locator, pair.locator)
+            genuine_pairs[ab] = genuine_pairs[ab[::-1]] = match
+            imposter_pairs[ab] = imposter_pairs[ab[::-1]] = not match
+
+    for ts, tsm in zip(test_samples, metrics):
+        for pair, tsm_pair in zip(ts.pairs, tsm.pair_samples):
+            ab = (ts.locator, pair.locator)
+            fm[ab] = fm[ab[::-1]] = tsm_pair.is_false_match
+            fnm[ab] = fnm[ab[::-1]] = tsm_pair.is_false_non_match
+            pair_failures[ab] = pair_failures[ab[::-1]] = tsm_pair.failure_to_enroll
+            fte[ab] = fte[ab[::-1]] = (
+                tsm.bbox_failure_to_enroll or tsm.keypoint_failure_to_align or tsm_pair.failure_to_enroll
+            )
+
+    n_genuine_pairs = np.sum([genuine_pairs[pair] for pair in unique_pairs])
+    n_imposter_pairs = np.sum([imposter_pairs[pair] for pair in unique_pairs])
+    n_fm = np.sum([fm[pair] for pair in unique_pairs])
+    n_fnm = np.sum([fnm[pair] for pair in unique_pairs])
+    n_pair_failures = np.sum([pair_failures[pair] for pair in unique_pairs])
+    n_fte = np.sum([fte[pair] for pair in unique_pairs])
+
+    return (n_genuine_pairs, n_imposter_pairs, n_fm, n_fnm, n_pair_failures, n_fte)
```

Comment:
FYI when there are many returned values like here it's better to create an object.

Reply:
And then you can include things like FNMR, FMR and FTER as well.